lpeg/modules/lpeg/printf.lua

```diff
@@ -0,0 +1,207 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+--[[
+# printf Parser Module
+
+See: http://www.cplusplus.com/reference/cstdio/printf/
+
+## Variables
+- none
+
+## Functions
+
+### build_printf_grammar
+
+Constructs an LPEG grammar based on the printf format/args array.
+
+*Arguments*
+- t (array) - Column one is the format string and the additional columns are the
+  capture names of the arguments. Capture names starting with an `@` are lpeg.re
+  expressions so the default parser for a type can be overridden when necessary.
+
+*Return*
+- grammar (LPEG user data object) or an error is thrown
+--]]
+
+-- Imports
+local string    = require "string"
+local l         = require "lpeg"
+l.locale(l)
+local es        = require "lpeg.escape_sequences"
+local re        = require "re"
+
+local ipairs    = ipairs
+local error     = error
+local type      = type
+local tonumber  = tonumber
+
+local M = {}
+setfenv(1, M) -- Remove external access to contain everything in the module
+
+local flags         = l.Cg((l.S"-+ #0")^0, "flags")
+local width         = l.Cg(l.digit^1 + "*", "width")
+local precision     = l.P"." * l.Cg(l.digit^0 + l.P"*" + l.Cc"0", "precision")
+local length        = l.Cg(l.P"hh" + "h" + "ll" + "l" + "j" + "z" + "t" + "L", "length")
+local specifier     = l.Cg(l.S"diuoxXfFeEgGaAcspn%", "specifier")
+local format        = l.Ct(l.P"%" * flags^-1 * width^-1 * precision^-1 * length^-1 * specifier)
+local literal_c     = 1 - l.S"%\\"
+local literal       = l.Cs((literal_c + es.c + l.P"%%"/"%%")^1)
+local segment       = literal + format
+local fmt_string    = l.Ct(segment * segment^0)
+
+local sign      = l.S"+-"
+local decimal   = l.P"." * l.digit^0
+local exponent  = l.S"eE" * sign^-1 * l.digit^1
+local float     = sign^-1 * l.digit^1 * decimal^-1 * exponent^-1
+local hdecimal  = l.P"." * l.xdigit^0
+local hexponent = l.S"pP" * sign^-1 * l.digit^1
+local hfloat    = sign^-1 * l.P"0" * l.S"xX" * l.xdigit^1 * hdecimal^-1 * hexponent^-1
+local sp        = l.P" "
+
```

Comment:
It doesn't look like string precision to limit output is handled, for example a test case like:

`{{"%.3s", "s"}, "AAABBB", {s = "AAA"}}`

or

`{{"%.3s%.3s", "s", "t"}, "AAABBB", {s = "AAA", t = "BBB"}}`

Reply:
Fixing: also the first example should actually be a no match